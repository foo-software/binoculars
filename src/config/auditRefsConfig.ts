export const seoAuditRefs = [
  // custom Binoculars content group
  {
    id: 'title-length',
    group: 'meaningful-content',
    weight: 7,
  },
  {
    id: 'description-length',
    group: 'meaningful-content',
    weight: 7,
  },
  {
    id: 'headings',
    group: 'meaningful-content',
    weight: 5,
  },
  {
    id: 'meaningful-text',
    group: 'meaningful-content',
    weight: 5,
  },
  {
    id: 'meaningful-tag-structure',
    group: 'meaningful-content',
    weight: 5,
  },
  {
    id: 'keywords',
    group: 'meaningful-content',
    weight: 7,
  },
  // from SEO category
  {
    id: 'viewport',
    weight: 5,
    group: 'seo-mobile',
  },
  {
    id: 'document-title',
    weight: 10,
    group: 'seo-content',
  },
  {
    id: 'meta-description',
    weight: 10,
    group: 'seo-content',
  },
  {
    id: 'http-status-code',
    weight: 1,
    group: 'seo-crawl',
  },
  {
    id: 'link-text',
    weight: 2,
    group: 'seo-content',
  },
  {
    id: 'crawlable-anchors',
    weight: 1,
    group: 'seo-crawl',
  },
  {
    id: 'is-crawlable',
    weight: 5,
    group: 'seo-crawl',
  },
  {
    id: 'robots-txt',
    weight: 2,
    group: 'seo-crawl',
  },
  {
    id: 'image-alt',
    weight: 1,
    group: 'seo-content',
  },
  {
    id: 'hreflang',
    weight: 2,
    group: 'seo-content',
  },
  {
    id: 'canonical',
    weight: 1,
    group: 'seo-content',
  },
  {
    id: 'font-size',
    weight: 1,
    group: 'seo-mobile',
  },
  {
    id: 'plugins',
    weight: 1,
    group: 'seo-content',
  },
  {
    id: 'tap-targets',
    weight: 1,
    group: 'seo-mobile',
  },
  {
    id: 'structured-data',
    weight: 0,
  },
  // from accessibility category
  {
    id: 'accesskeys',
    weight: 0,
    group: 'a11y-navigation',
  },
  {
    id: 'aria-allowed-attr',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'aria-hidden-body',
    weight: 5,
    group: 'a11y-aria',
  },
  {
    id: 'aria-hidden-focus',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'aria-input-field-name',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'aria-required-attr',
    weight: 10,
    group: 'a11y-aria',
  },
  {
    id: 'aria-required-children',
    weight: 5,
    group: 'a11y-aria',
  },
  {
    id: 'aria-required-parent',
    weight: 5,
    group: 'a11y-aria',
  },
  {
    id: 'aria-roles',
    weight: 5,
    group: 'a11y-aria',
  },
  {
    id: 'aria-toggle-field-name',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'aria-valid-attr-value',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'aria-valid-attr',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'button-name',
    weight: 0,
    group: 'a11y-names-labels',
  },
  {
    id: 'bypass',
    weight: 3,
    group: 'a11y-navigation',
  },
  {
    id: 'color-contrast',
    weight: 3,
    group: 'a11y-color-contrast',
  },
  {
    id: 'definition-list',
    weight: 0,
    group: 'a11y-tables-lists',
  },
  {
    id: 'dlitem',
    weight: 0,
    group: 'a11y-tables-lists',
  },
  {
    id: 'duplicate-id-active',
    weight: 0,
    group: 'a11y-navigation',
  },
  {
    id: 'duplicate-id-aria',
    weight: 0,
    group: 'a11y-aria',
  },
  {
    id: 'form-field-multiple-labels',
    weight: 0,
    group: 'a11y-names-labels',
  },
  {
    id: 'frame-title',
    weight: 0,
    group: 'a11y-names-labels',
  },
  {
    id: 'heading-order',
    weight: 2,
    group: 'a11y-navigation',
  },
  {
    id: 'html-has-lang',
    weight: 3,
    group: 'a11y-language',
  },
  {
    id: 'html-lang-valid',
    weight: 3,
    group: 'a11y-language',
  },
  {
    id: 'image-alt',
    weight: 10,
    group: 'a11y-names-labels',
  },
  {
    id: 'input-image-alt',
    weight: 0,
    group: 'a11y-names-labels',
  },
  {
    id: 'label',
    weight: 0,
    group: 'a11y-names-labels',
  },
  {
    id: 'layout-table',
    weight: 0,
    group: 'a11y-tables-lists',
  },
  {
    id: 'link-name',
    weight: 3,
    group: 'a11y-names-labels',
  },
  {
    id: 'list',
    weight: 3,
    group: 'a11y-tables-lists',
  },
  {
    id: 'listitem',
    weight: 3,
    group: 'a11y-tables-lists',
  },
  {
    id: 'meta-refresh',
    weight: 0,
    group: 'a11y-best-practices',
  },
  {
    id: 'meta-viewport',
    weight: 3,
    group: 'a11y-best-practices',
  },
  {
    id: 'object-alt',
    weight: 0,
    group: 'a11y-names-labels',
  },
  {
    id: 'tabindex',
    weight: 0,
    group: 'a11y-navigation',
  },
  {
    id: 'td-headers-attr',
    weight: 0,
    group: 'a11y-tables-lists',
  },
  {
    id: 'th-has-data-cells',
    weight: 0,
    group: 'a11y-tables-lists',
  },
  {
    id: 'valid-lang',
    weight: 0,
    group: 'a11y-language',
  },
  {
    id: 'video-caption',
    weight: 0,
    group: 'a11y-audio-video',
  },
  {
    id: 'video-description',
    weight: 0,
    group: 'a11y-audio-video',
  },
  {
    id: 'logical-tab-order',
    weight: 0,
  },
  {
    id: 'focusable-controls',
    weight: 0,
  },
  {
    id: 'interactive-element-affordance',
    weight: 0,
  },
  {
    id: 'managed-focus',
    weight: 0,
  },
  {
    id: 'focus-traps',
    weight: 0,
  },
  {
    id: 'custom-controls-labels',
    weight: 0,
  },
  {
    id: 'custom-controls-roles',
    weight: 0,
  },
  {
    id: 'visual-order-follows-dom',
    weight: 0,
  },
  {
    id: 'offscreen-content-hidden',
    weight: 0,
  },
  {
    id: 'use-landmarks',
    weight: 0,
  },
  // misc
  {
    id: 'final-screenshot',
    weight: 0,
    group: 'seo-content',
  },
];
