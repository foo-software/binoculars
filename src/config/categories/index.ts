export { default as getSeo } from './seo';
