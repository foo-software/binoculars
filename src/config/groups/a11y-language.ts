import defaultConfig from 'lighthouse/lighthouse-core/config/default-config';
import { getFormatted } from 'lighthouse/lighthouse-core/lib/i18n/i18n';
import replaceLinks from '../../helpers/replaceLinks';

export default (locale: string | undefined | unknown = 'en') => ({
  description: replaceLinks(
    getFormatted(defaultConfig.groups['a11y-language'].description, locale),
  ),
  title: replaceLinks(
    getFormatted(defaultConfig.groups['a11y-language'].title, locale),
  ),
});
