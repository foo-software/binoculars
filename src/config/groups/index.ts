export { default as getA11yAria } from './a11y-aria';
export { default as getA11yAudioVideo } from './a11y-audio-video';
export { default as getA11yBestPractices } from './a11y-best-practices';
export { default as getA11yColorContrast } from './a11y-color-contrast';
export { default as getA11yLanguage } from './a11y-language';
export { default as getA11yNamesLabels } from './a11y-names-labels';
export { default as getA11yNavigation } from './a11y-navigation';
export { default as getA11yTablesLists } from './a11y-tables-lists';
export { default as getSeoContent } from './seo-content';
export { default as getSeoCrawl } from './seo-crawl';
export { default as getSeoMobile } from './seo-mobile';
