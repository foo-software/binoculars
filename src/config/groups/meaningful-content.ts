// @TODO support a `locale` argument and translation
export default () => ({
  description:
    'Use keywords and semantic HTML to establish meaningful content.',
  title: 'Meaningful Content',
});
