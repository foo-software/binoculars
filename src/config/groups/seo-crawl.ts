import { getFormatted } from 'lighthouse/lighthouse-core/lib/i18n/i18n';
import defaultConfig from 'lighthouse/lighthouse-core/config/default-config';
import replaceLinks from '../../helpers/replaceLinks';

export default (locale: string | undefined | unknown = 'en') => ({
  description: replaceLinks(
    getFormatted(defaultConfig.groups['seo-crawl'].description, locale),
  ),
  title: replaceLinks(
    getFormatted(defaultConfig.groups['seo-crawl'].title, locale),
  ),
});
