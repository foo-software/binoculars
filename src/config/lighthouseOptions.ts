export const PORT = process.env.BINOCULARS_CHROME_PORT || 4000;

export default {
  chromeFlags: [
    '--disable-dev-shm-usage',
    '--headless',
    '--no-sandbox',
    '--ignore-certificate-errors',
  ],
  port: PORT,
};
