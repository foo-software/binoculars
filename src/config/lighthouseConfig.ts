import { seoAuditRefs } from './auditRefsConfig';
import a11yAria from './groups/a11y-aria';
import a11yAudioVideo from './groups/a11y-audio-video';
import a11yBestPractices from './groups/a11y-best-practices';
import a11yColorContrast from './groups/a11y-color-contrast';
import a11yLanguage from './groups/a11y-language';
import a11yNamesLabels from './groups/a11y-names-labels';
import a11yNavigation from './groups/a11y-navigation';
import a11yTablesLists from './groups/a11y-tables-lists';
import DescriptionLengthAudit from '../audits/seo/description-length';
import meaningfulContent from './groups/meaningful-content';
import seo from './categories/seo';
import seoContent from './groups/seo-content';
import seoCrawl from './groups/seo-crawl';
import seoMobile from './groups/seo-mobile';
import TitleElementGatherer from '../gatherers/title-element';
import TitleLengthAudit from '../audits/seo/title-length';

// much of this was taken from
// https://github.com/GoogleChrome/lighthouse/blob/b834427d676dc77e112d124ca42cc588f896950e/lighthouse-core/config/constants.js#L8-L41
/**
 * Adjustments needed for DevTools network throttling to simulate
 * more realistic network conditions.
 * @see https://crbug.com/721112
 * @see https://docs.google.com/document/d/10lfVdS1iDWCRKQXPfbxEn4Or99D64mvNlugP1AQuFlE/edit
 */
const DEVTOOLS_RTT_ADJUSTMENT_FACTOR = 3.75;
const DEVTOOLS_THROUGHPUT_ADJUSTMENT_FACTOR = 0.9;

const throttling = {
  DEVTOOLS_RTT_ADJUSTMENT_FACTOR,
  DEVTOOLS_THROUGHPUT_ADJUSTMENT_FACTOR,
  // These values align with WebPageTest's definition of 'Fast 3G'
  // But offer similar charateristics to roughly the 75th percentile of 4G connections.
  mobileSlow4G: {
    rttMs: 150,
    throughputKbps: 1.6 * 1024,
    requestLatencyMs: 150 * DEVTOOLS_RTT_ADJUSTMENT_FACTOR,
    downloadThroughputKbps: 1.6 * 1024 * DEVTOOLS_THROUGHPUT_ADJUSTMENT_FACTOR,
    uploadThroughputKbps: 750 * DEVTOOLS_THROUGHPUT_ADJUSTMENT_FACTOR,
    cpuSlowdownMultiplier: 4,
  },
  // These values partially align with WebPageTest's definition of 'Regular 3G'.
  // These values are meant to roughly align with Chrome UX report's 3G definition which are based
  // on HTTP RTT of 300-1400ms and downlink throughput of <700kbps.
  mobileRegluar3G: {
    rttMs: 300,
    throughputKbps: 700,
    requestLatencyMs: 300 * DEVTOOLS_RTT_ADJUSTMENT_FACTOR,
    downloadThroughputKbps: 700 * DEVTOOLS_THROUGHPUT_ADJUSTMENT_FACTOR,
    uploadThroughputKbps: 700 * DEVTOOLS_THROUGHPUT_ADJUSTMENT_FACTOR,
    cpuSlowdownMultiplier: 4,
  },
  // Using a 'broadband' connection type
  // Corresponds to 'Dense 4G 25th percentile' in https://docs.google.com/document/d/1Ft1Bnq9-t4jK5egLSOc28IL4TvR-Tt0se_1faTA4KTY/edit#heading=h.bb7nfy2x9e5v
  desktopDense4G: {
    rttMs: 40,
    throughputKbps: 10 * 1024,
    cpuSlowdownMultiplier: 1,
    requestLatencyMs: 0, // 0 means unset
    downloadThroughputKbps: 0,
    uploadThroughputKbps: 0,
  },
};

// https://github.com/GoogleChrome/lighthouse/blob/master/docs/configuration.md
// okay, maybe a little different from...
// https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/config/perf-config.js
export default (locale: string | undefined | unknown) => ({
  extends: 'lighthouse:default',
  audits: [DescriptionLengthAudit, TitleLengthAudit],
  categories: {
    seo: {
      ...seo(locale),
      auditRefs: seoAuditRefs,
    },
  },
  groups: {
    'a11y-aria': a11yAria(locale),
    'a11y-audio-video': a11yAudioVideo(locale),
    'a11y-best-practices': a11yBestPractices(locale),
    'a11y-color-contrast': a11yColorContrast(locale),
    'a11y-language': a11yLanguage(locale),
    'a11y-names-labels': a11yNamesLabels(locale),
    'a11y-navigation': a11yNavigation(locale),
    'a11y-tables-lists': a11yTablesLists(locale),
    'meaningful-content': meaningfulContent(),
    'seo-content': seoContent(locale),
    'seo-crawl': seoCrawl(locale),
    'seo-mobile': seoMobile(locale),
  },
  passes: [
    {
      passName: 'defaultPass',
      gatherers: [TitleElementGatherer],
    },
  ],
  settings: {
    emulatedFormFactor: 'mobile',
    onlyCategories: ['seo'],
    throttling: throttling.mobileSlow4G,
    throttlingMethod: 'simulate',
  },
});
