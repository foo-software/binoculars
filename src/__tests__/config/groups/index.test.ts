import * as groups from '../../../config/groups';

interface DefinitionCollection {
  [key: string]: any;
}

const auditCollection: DefinitionCollection = groups;

describe('groups config', () => {
  for (const key in auditCollection) {
    describe(`${key}()`, () => {
      const audit = auditCollection[key]();
      it('should match snapshot', () => {
        expect(audit).toMatchSnapshot();
      });
    });
  }
});
