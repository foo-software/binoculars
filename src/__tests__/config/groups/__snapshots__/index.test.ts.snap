// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groups config getA11yAria() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to improve the usage of ARIA in your application which may enhance the experience for users of assistive technology, like a screen reader.",
  "title": "ARIA",
}
`;

exports[`groups config getA11yAudioVideo() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to provide alternative content for audio and video. This may improve the experience for users with hearing or vision impairments.",
  "title": "Audio and video",
}
`;

exports[`groups config getA11yBestPractices() should match snapshot 1`] = `
Object {
  "description": "These items highlight common accessibility best practices.",
  "title": "Best practices",
}
`;

exports[`groups config getA11yColorContrast() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to improve the legibility of your content.",
  "title": "Contrast",
}
`;

exports[`groups config getA11yLanguage() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to improve the interpretation of your content by users in different locales.",
  "title": "Internationalization and localization",
}
`;

exports[`groups config getA11yNamesLabels() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to improve the semantics of the controls in your application. This may enhance the experience for users of assistive technology, like a screen reader.",
  "title": "Names and labels",
}
`;

exports[`groups config getA11yNavigation() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to improve keyboard navigation in your application.",
  "title": "Navigation",
}
`;

exports[`groups config getA11yTablesLists() should match snapshot 1`] = `
Object {
  "description": "These are opportunities to improve the experience of reading tabular or list data using assistive technology, like a screen reader.",
  "title": "Tables and lists",
}
`;

exports[`groups config getSeoContent() should match snapshot 1`] = `
Object {
  "description": "Format your HTML in a way that enables crawlers to better understand your app’s content.",
  "title": "Content Best Practices",
}
`;

exports[`groups config getSeoCrawl() should match snapshot 1`] = `
Object {
  "description": "To appear in search results, crawlers need access to your app.",
  "title": "Crawling and Indexing",
}
`;

exports[`groups config getSeoMobile() should match snapshot 1`] = `
Object {
  "description": "Make sure your pages are mobile friendly so users don’t have to pinch or zoom in order to read the content pages. [Learn more](https://developers.google.com/search/mobile-sites/).",
  "title": "Mobile Friendly",
}
`;
